src/core/mac/sub_mac.cpp

```diff
@@ -227,6 +240,32 @@ void SubMac::StartCsmaBackoff(void)
     uint32_t backoff;
     uint32_t backoffExponent = kMinBE + mTransmitRetries + mCsmaBackoffs;
 
+#if OPENTHREAD_CONFIG_CSL_TRANSMITTER_ENABLE
+    if (mTransmitFrame.mInfo.mTxInfo.mPeriod != 0)
+    {
+        uint32_t phaseNow     = (otPlatTimeGet() / OT_US_PER_TEN_SYMBOLS) % mTransmitFrame.mInfo.mTxInfo.mPeriod;
+        uint32_t phaseDesired = mTransmitFrame.mInfo.mTxInfo.mPhase;
+
+        mTransmitFrame.SetCsmaCaEnabled(false);
+
+        SetState(kStateCslTransmit);
+
+        if (phaseNow < phaseDesired)
+        {
+            mTimer.Start((phaseDesired - phaseNow) * OT_US_PER_TEN_SYMBOLS);
+        }
+        else if (phaseNow > phaseDesired)
+        {
+            mTimer.Start((mTransmitFrame.mInfo.mTxInfo.mPeriod + phaseDesired - phaseNow) * OT_US_PER_TEN_SYMBOLS);
```

Comment:
Rather than holding the transmission queue for another CSL period, how about just call `HandleTransmitDone` with some error number to indicate the desired transmission time is in the past. MAC should reschedule this CSL frame next time when appropriate?

Reply:
Well, I prefer not to call `HandleTransmitDone` here. Rescheduling the transmission in `HandleTransmitDone` makes this process look like a transmission retry, but it's actually not a retry. And `HandleTransmitDone`actually handles retry.